File: app/src/main/java/cz/civilizacehra/cipherbreaker/PrincipTrainerActivity.java
package cz.civilizacehra.cipherbreaker;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

public class PrincipTrainerActivity extends Activity {

    ImageView principy[] = new ImageView[6];
    Switch invertSwitch;

    ImageView imageView;
    ImageView solutions[] = new ImageView[5];

    TextView textView;
    Button nextButton;
    Button solutionButton;

    int mHihglightColor = 0xFF24872F;
    int mWrongColor = 0xFFBF3434;
    int mDefaultColor = 0xFFBBBBBB;

    boolean mInvert;
    String mPrincipy[] = {"n", "m", "b", "s", "bin", "t"};
    String mPrincip = "n";
    Random mRandom = new Random();
    int mCurrent = 0, mLast = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_princip_trainer);

        principy[0] = (ImageView)findViewById(R.id.nImage);
        principy[1] = (ImageView)findViewById(R.id.mImage);
        principy[2] = (ImageView)findViewById(R.id.bImage);
        principy[3] = (ImageView)findViewById(R.id.sImage);
        principy[4] = (ImageView)findViewById(R.id.binImage);
        principy[5] = (ImageView)findViewById(R.id.tImage);

        for (int i = 0; i < mPrincipy.length; ++i) {
            final int ii = i;
            if (i == 0) {
                principy[i].setBackgroundColor(mHihglightColor);
            } else {
                principy[i].setBackgroundColor(mDefaultColor);
            }
            principy[i].setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    mPrincip = mPrincipy[ii];
                    for (int i = 0; i < mPrincipy.length; ++i) {
                        if (i == ii) {
                            principy[i].setBackgroundColor(mHihglightColor);
                        } else {
                            principy[i].setBackgroundColor(mDefaultColor);
                        }
                    }
                    newLetter();
                }
            });
        }

        invertSwitch = (Switch)findViewById(R.id.invertSwitch);
        invertSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                mInvert = isChecked;
                newLetter();

            }
        });

        imageView = (ImageView)findViewById(R.id.imageView);

        solutions[0] = (ImageView)findViewById(R.id.solution1);
        solutions[1] = (ImageView)findViewById(R.id.solution2);
        solutions[2] = (ImageView)findViewById(R.id.solution3);
        solutions[3] = (ImageView)findViewById(R.id.solution4);
        solutions[4] = (ImageView)findViewById(R.id.solution5);

        for (int i = 0; i < 5; ++i) {
            final int ii = i;
            solutions[ii].setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    if (ii == mCurrent){
                        new Thread(new Runnable() {
                            @Override
                            public void run() {
                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        solutions[ii].setBackgroundColor(mHihglightColor);
                                        textView.setTextColor(mHihglightColor);
                                        textView.setText("Correct !");
                                    }
                                });
                                try {
                                    Thread.sleep(300);
                                } catch (InterruptedException e) {
                                    e.printStackTrace();
                                }
                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        newLetter();
                                    }
                                });
                            }
                        }).start();
                    } else {
                        solutions[ii].setBackgroundColor(mWrongColor);
                        textView.setTextColor(mWrongColor);
                        textView.setText("Wrong !!!");
                    }
                }
            });
        }

        textView = (TextView)findViewById(R.id.textView);

        nextButton = (Button)findViewById(R.id.nextButton);
        nextButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                newLetter();
            }
        });

        solutionButton = (Button)findViewById(R.id.solutionButton);
        solutionButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                solutions[mCurrent].setBackgroundColor(mHihglightColor);
                /*textView.setTextColor(mHihglightColor);
                textView.setText("Correct !");*/
            }
        });

        setImage(principy[0], "principy/n/7.png");
        setImage(principy[1], "principy/m/16.png");
        setImage(principy[2], "principy/b/9.png");
        setImage(principy[3], "principy/s/13.png");
        setImage(principy[4], "principy/bin/21.png");
        setImage(principy[5], "principy/t/15.png");

        newLetter();
    }

    private void setImage(ImageView btn, String path) {
        InputStream is = null;
        try {
            is = this.getResources().getAssets().open(path);
        } catch (IOException e) {
            Log.w("EL", e);
        }

        Bitmap image = BitmapFactory.decodeStream(is);
        btn.setImageBitmap(image);
        imageView.setBackgroundColor(mDefaultColor);
    }

    private void newLetter() {
        for (int i = 0; i < 5; ++i) {
            solutions[i].setBackgroundColor(mDefaultColor);
            textView.setText("");
        }

        String mSource;
        String mOut;
        mCurrent = mRandom.nextInt(5);

        if (mInvert) {
            mSource = mPrincip;
            mOut = "a";
        } else {
            mSource = "a";
            mOut = mPrincip;
        }

        ArrayList<Integer> list = new ArrayList<>();
        for (int i=1; i<=26; i++) {
            if (i != mLast) {
                list.add(i);
            }
        }
        Collections.shuffle(list);
        for (int i=0; i<5; i++) {
            setImage(solutions[i], "principy/" + mSource + "/" + list.get(i) + ".png");
        }
        mLast = list.get(mCurrent);

        InputStream is = null;
        try {
            is = this.getResources().getAssets().open("principy/" + mOut + "/" + list.get(mCurrent) + ".png");
        } catch (IOException e) {
            Log.w("EL", e);
        }

        Bitmap image = BitmapFactory.decodeStream(is);
        imageView.setImageBitmap(image);
        imageView.setBackgroundColor(mDefaultColor);
    }
}

File: app/src/main/java/cz/civilizacehra/cipherbreaker/PrincipTrainerActivity.kt
package cz.civilizacehra.cipherbreaker

import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Log
import android.widget.ImageView
import android.widget.Switch

import java.io.IOException
import java.io.InputStream
import java.util.ArrayList

class PrincipTrainerActivity : BasePrincipActivity() {

    private val imageView by lazy { findViewById<ImageView>(R.id.imageView) }
    private val solutions by lazy { arrayOf<ImageView>(
            findViewById(R.id.solution1),
            findViewById(R.id.solution2),
            findViewById(R.id.solution3),
            findViewById(R.id.solution4),
            findViewById(R.id.solution5)
    ) }

    private var mInvert: Boolean = false
    private var mCurrent = 0
    private var mLast = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_princip_trainer)
        super.onCreate(savedInstanceState)

        findViewById<Switch>(R.id.invertSwitch).setOnCheckedChangeListener { _, isChecked ->
            mInvert = isChecked
            solutionView.text = ""
            newExample()
        }

        for (i in 0..4) {
            solutions[i].setOnClickListener {
                if (i == mCurrent) {
                    Thread(Runnable {
                        runOnUiThread {
                            setSolution(i, "Correct !", mHihglightColor)
                        }
                        try {
                            Thread.sleep(300)
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }

                        runOnUiThread {
                            solutionView.text = ""
                            newExample()
                        }
                    }).start()
                } else {
                    setSolution(i, "Wrong !!!", mWrongColor)
                }
            }
        }

        solutionButton.setOnClickListener {
            solutions[mCurrent].setBackgroundColor(mHihglightColor)
        }

        newExample()
    }

    private fun setSolution(index: Int, text: String, color: Int) {
        solutions[index].setBackgroundColor(color)
        solutionView.setTextColor(color)
        solutionView.text = text
    }

    override fun newExample() {
        for (i in 0..4) {
            solutions[i].setBackgroundColor(mDefaultColor)
        }

        mCurrent = mRandom.nextInt(5)

        val mSource = if (mInvert) mPrincip else "a"
        val mOut = if (mInvert) "a" else mPrincip

        val list = ArrayList<Int>()
        for (i in 1..26) {
            if (i != mLast) {
                list.add(i)
            }
        }
        list.shuffle()
        for (i in 0..4) {
            setImage(solutions[i], "principy/$mSource/${list[i]}.png")
        }
        mLast = list[mCurrent]

        var `is`: InputStream? = null
        try {
            `is` = this.resources.assets.open("principy/$mOut/${list[mCurrent]}.png")
        } catch (e: IOException) {
            Log.w("EL", e)
        }

        val image = BitmapFactory.decodeStream(`is`)
        imageView.setImageBitmap(image)
        imageView.setBackgroundColor(mDefaultColor)
    }
}
